### ISSUE TITLE
Improve custom configuration variables

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode-cpptools#6215]
@sean-mcmanus I think my comment is already a suggestion to make it better:
1) Allow <code>customConfigurationVariables</code> to have property values which are arrays of strings;
2) Add a new class of VS Code variables which can be used in <code>tasks.json</code> and <code>launch.json</code>. These variables can be used in substitutions with a syntax like <code>${cpptools:<Variablename\>}</code>. This is similar to the syntax for [environment variables](https://code.visualstudio.com/docs/editor/variables-reference#_environment-variables), [configuration variables](https://code.visualstudio.com/docs/editor/variables-reference#_configuration-variables), and [command variables](https://code.visualstudio.com/docs/editor/variables-reference#_command-variables). These substitutions should also be allowed as part of an array of strings similar to the [example](https://code.visualstudio.com/docs/cpp/c-cpp-properties-schema-reference#_example) for <code>env</code> variablles in <code>c_cpp_properties.json</code> files. So in arrays of strings the variable can not only represent a string but also an array of strings similar to the following lines in the mentioned example: 
```"env": { "myDefaultIncludePath": ["${workspaceFolder}", "${workspaceFolder}/include"] },```
```"includePath": ["${myDefaultIncludePath}", "/another/path"]```
which is substituted to
```"includePath": ["${workspaceFolder}", "${workspaceFolder}/include", "/another/path"]```

_Originally posted by @PetaSoft in https://github.com/microsoft/vscode-cpptools/pull/5453#issuecomment-696612225_

**Comment by bobbrow** [comment]
I don't think that `${cpptools:<Variablename>}` is possible.  Commands need to be registered before they can be invoked.  It is not dynamic as far as I am aware.  I don't think there is a way for VS Code to say "Hey cpptools, do you know what to do with `<VariableName>`?"  I believe this would be a feature ask for the VS Code team.

**Comment by bobbrow** [comment]
But thinking about the scenario in (2) a little more, perhaps we can achieve something similar with our new build tasks support coming in 1.1.  When a task type is 'cppbuild' we'll get a chance to intercept the commands before they are executed and we could do a substitution there.  I'm not sure if we get the same chance with the launch.json.

**Comment by spacemonkd** [comment]
Can I try and work on this? @sean-mcmanus sir @bobbrow sir

**Comment by sean-mcmanus** [comment]
@devabhishekpal Seems fine to me, but I think you'd need to use the recent task changes in the master branch, but our published cpptools binary has been updated so let us know if you hit any strange issues due to that and we could publish an update with our pending 1.1.0-insiders binaries (I tried it out with the published binaries and I didn't see any issues though).

**Comment by spacemonkd** [comment]
Okay sir I am checking

**Comment by spacemonkd** [comment]
Sorry I wasnt able to work on this for a while due to network issues. I am starting work from today @sean-mcmanus sir

**Comment by spacemonkd** [comment]
Could you point me to the file where I could start? @sean-mcmanus sir

**Comment by spacemonkd** [comment]
I am really extremely sorry for being this late. We had exams and so I wasnt able to give my time.

**Comment by sean-mcmanus** [comment]
The previous PR in this area involved the files at https://github.com/microsoft/vscode-cpptools/pull/5453/files .

**Comment by spacemonkd** [comment]
Could you give me some suggestions as to what variables types to add? @sean-mcmanus sir @bobbrow sir. 
Because in tasks.json I cannot find a possible use case for string arrays except in case of the `args` property and `dependsOn` property in which we could use a string array with. But other than that could you give me some examples of replacements of such properties? It would be really helpful.

**Comment by sean-mcmanus** [comment]
I think the desire for a array of strings was in regards to the array of strings in c_cpp_properties, e.g. https://github.com/microsoft/vscode-cpptools/blob/master/Extension/c_cpp_properties.schema.json#L23 and not tasks.json (other than args maybe).

**Comment by TDHolmes** [comment]
Can this enable configuration variables that read from the user's terminal environment? I was confused to find that [other parts of vscode](https://code.visualstudio.com/docs/editor/variables-reference#_environment-variables) support reading from users environment variables, but this extension instead uses that syntax (`{env:VAR_NAME}`) to read from a [hard-coded dictionary of configuration values](https://code.visualstudio.com/docs/cpp/c-cpp-properties-schema-reference#_toplevel-properties).

**Comment by akbyrd** [comment]
Related: I've made a request for vscode to support passing arguments to tasks. This extension uses the "recommended" pattern of stashing global context somewhere, then retrieving it through a command input variable inside a task. Ideally this could be streamlined so that global context + a command + an extension are not required to pass an input value to a task.
https://github.com/microsoft/vscode/issues/178343

I'm not 100% sure whether that would help this exact case of passing information about the current C++ configuration to a task, but there's another gap where it might. Right now we need separate tasks for "build", "clean", and "rebuild". And also for building a file, project, or the entire solution. In many build systems these are nearly the same invocation, but with a single word difference. And as a user I might want each of those to reuse the same task. This allows them to share a task panel dedicated to builds. It also allows them to share the same problem matcher source so problems are always up to date with the latest build.

The pattern used by this extension works "ok" for the current case of `customConfigurationVariables`, but it's not very robust. That's not a shortcoming of this extension, just a gap in vscode. If the input was something intended for the current task only and wasn't effectively immutable it becomes quite hard to ensure correctness. What if multiple tasks happen to run that use the same variable name? If I want to make a task that builds both Project A and Project B how do I get the project name into each invocation of the build task without stomping over input to the other task?

To make things a little more concrete here is one of my current build tasks
```
"inputs": [
	{
		"id": "solution",
		"type": "command",
		"command": "cpptools.activeConfigCustomVariable",
		"args": "solution",
	},
	{
		"id": "configuration",
		"type": "command",
		"command": "cpptools.activeConfigCustomVariable",
		"args": "configuration",
	},
	{
		"id": "platform",
		"type": "command",
		"command": "cpptools.activeConfigCustomVariable",
		"args": "platform",
	},
	{
		"id": "target",
		"type": "command",
		"command": "akbyrd.task.getArgs",
		"args": "target",
	},
],
"tasks": [
	{
		"label": "build",
		"icon": { "id": "code", "color": "terminal.ansiCyan" },
		"detail": "msbuild",
		"group": "build",
		"type": "cppbuild",
		"command": "msbuild",
		"options": { "cwd": "${workspaceFolder}" },
		"args": [
			"/nologo", "/v:m", "/m",
			"/t:${input:target}",
			"/p:Configuration=${input:configuration}",
			"/p:Platform=${input:platform}",
			"${input:solution}",
		],
		"problemMatcher": "$akbyrd-msbuild",
		"presentation": {
			"panel": "dedicated",
			"reveal": "silent",
			"clear": true,
			"showReuseMessage": false,
			"revealProblems": "onProblem",
		},
	},
],
```

That I invoke like so
```
{ "key": "ctrl+alt+u", "command": "akbyrd.task.runWithArgs", "args": { "task": "build", "taskArgs": { "target": "Build",   } },
{ "key": "ctrl+alt+h", "command": "akbyrd.task.runWithArgs", "args": { "task": "build", "taskArgs": { "target": "Rebuild", } },
{ "key": "ctrl+alt+b", "command": "akbyrd.task.runWithArgs", "args": { "task": "build", "taskArgs": { "target": "Clean",   } },
